### DISCUSSION TITLE
How to integrate my own robot

### DISCUSSION THREAD

**Discussion by savolla** [tesseract-robotics/tesseract_ros#148] [Q&A] [answered]
Hi, I want to migrate from Moveit to tesseract. I have 8 DOF welding robot and working on a project just like [this](https://www.youtube.com/watch?v=ZBk4dZvq0T4). I want to apply Trajopt since it provides nice cartesian trajectories. I have some questions;

1. How good is Trajopt compared to Descartes. Since I want to use mostly cartesian motions and avoid collisions. I watched [this](https://www.youtube.com/watch?v=AgW0Jtbq2xY) conference and saw you guys merge trajopt with descartes in one of the examples you show.
2. I know this library is ROS agnostic but which ROS distribution is the best for tesseract and trajopt?
3. I couldn't see an interface which provides an *easy* way to integrate my robot. I suppose I need to hack the examples for example basic_cartesian_example.cpp and it's basic_cartesian_example_node.cpp and its headers and the launch file and all bounded CMakeList.txt files which is very cumbersome and time consuming. Maybe I'm missing some details here.
4. I want to use Python instead of C++ with tesseract. I know there is a tesseract_python repo but when I clone and build it I cannot launch any of the examples because "I think" it doesn't have ROS support (yet?). I could find my way from there if I spawned my robot on rviz using tesseract_python.

Thanks for your attention!

**Comment by Levi-Armstrong** [comment, marked as answer]
@savolla Thank you for reaching out. I will put together a detailed response tonight.

**Reply by Levi-Armstrong** [reply]
> How good is Trajopt compared to Descartes. Since I want to use mostly cartesian motions and avoid collisions. I watched [this](https://www.youtube.com/watch?v=AgW0Jtbq2xY) conference and saw you guys merge trajopt with descartes in one of the examples you show.

Descartes is a sample based planner so the quality of the trajectory in most cases is propositional to how finely you can sample the search space. In reality this is not possible especial in your case with a gantry system the search space grows significantly and you run out of resources quickly. Now some work has been done to leverage similar techniques to RRT  to mitigate the memory issues which is an open PR. 

In the case of TrajOpt it is a local optimal planner so it needs a decent initial guess. In some cases a single start location can work, but if you the trajectory needs to make a configuration change at some point through the trajectory this can be problematic. This is were I would use descartes with a coarse sampling of the space which should be quick and then use TrajOpt to refine things and handle higher order things like tolerances waypoints, improving collision distance, smoothing acceleration and jerk.

In most case like sanding, painting, welding, etc there is a lot opportunity for parallelization which is where Tesseract's process planning shines where you can create taskflows significantly improving motion planning times for large industrial problems.

> I know this library is ROS agnostic but which ROS distribution is the best for tesseract and trajopt?

The following repositories tesseract, tesseract_planning and tesseract_gui are ROS agnostic but this repository tesseract_ros provides the ROS interfaces. There is most two items the Environment Monitor which is similar to the MoveIt PlanningScene which allows you to have a single server running and then multiple clients which monitor this server and keep up to date. The second is the planning server which allows you to submit process planning request. If you take a look at these they are wrappers around the agnostic libraries and not much is actually in these two especially the planning server. There is currently not python wrappers for these but one could easily port these to python versus creating a wrappers which I would happily review a PR for these. Long story short I would use this library if using ROS.

> I couldn't see an interface which provides an easy way to integrate my robot. I suppose I need to hack the examples for example basic_cartesian_example.cpp and it's basic_cartesian_example_node.cpp and its headers and the launch file and all bounded CMakeList.txt files which is very cumbersome and time consuming. Maybe I'm missing some details here.

I recommend take a look at the [tesseract_ros_workcell](https://github.com/tesseract-robotics/tesseract_ros_workcell) repository for how to leverage the ROS interfaces. The tesseract_examples and tesseract_ros_examples are good to get familiar with how to use the c++ API.

> I want to use Python instead of C++ with tesseract. I know there is a tesseract_python repo but when I clone and build it I cannot launch any of the examples because "I think" it doesn't have ROS support (yet?). I could find my way from there if I spawned my robot on rviz using tesseract_python.

I think the biggest hurdle is that there is not an equivalent to the EnvironmentMonitor and conversion utils from ros messages to tesseract types ported to python. I believe if these things were ported then you should be able to do everything in python. You could get by with launching the c++ environment monitory and interacting with using messages but you would still need the conversion utils ported. @johnwason  maybe to provided more details on the python since he currently maintains tesseract_python and did the python wrappers.

**Reply by savolla** [reply]
I already looked at tesseract_ros_workcell before but I get a build error which I already mentioned in the issue [here](https://github.com/tesseract-robotics/tesseract_ros_workcell/issues/13). So I decided to hack tesseract_ros_examples instead :) But I'll look at this again today. Thank you so much for your attention!

I also will try to contribute once I get a good idea about the code base! :)

**Reply by Levi-Armstrong** [reply]
I also recommend leveraging the ROP Inverse kinematics plugin for your gantry system and for the robot the OPW Inverse kinematics. If the robot is not the correct type to use the OPW IK solver then generate a IKFast plugin. This will have a huge impact on the planning times.

**Reply by johnwason** [reply]
Hello @savolla,

The Python wrappers are a direct wrapping of the C++ API using SWIG. This means that the documentation for C++ is mostly the same as for Python. It also means that the memory lifecycle of the data is the same as C++. This can be a little tricky since Python objects may not be valid if they are references to C++ objects that are deleted. There are a few other corner cases where the API is a little different. Unfortunately I haven't had funding to document everything, so for now you can look at the unit tests and examples from the viewer:

https://github.com/tesseract-robotics/tesseract_python/tree/master/tesseract_python/tests

https://github.com/tesseract-robotics/tesseract_python/tree/master/tesseract_viewer_python/examples

We have not been using the Python interface with ROS so we don't have an accompanying tesseract_python_ros (tesseract_ros_python?) package. If you are interested in contributing this piece we would be interested.

There is a Conda package I have put together for using tesseract_python with conda-forge: https://github.com/johnwason/tesseract-robotics-superpack-feedstock/blob/master/TESSERACT_README.md

There is also the PyRI teach pendant we are working on that uses Tesseract for planning: https://github.com/pyri-project/pyri-core/blob/master/README.md

If you have more questions you can open a discussion on tesseract_python. I don't follow the other repositories so you will need to @ me so I am notified.
